crates/lib/src/compiler/libquil.rs

```diff
@@ -0,0 +1,349 @@
+//! Uses [`libquil-sys`] as the client for Quilc
+
+use std::ffi::NulError;
+use std::num::TryFromIntError;
+use std::str::FromStr;
+use std::string::String;
+use std::{convert::TryFrom, ffi::CString};
+
+use super::quilc::{self, NativeQuilMetadata};
+
+/// The errors that can arise when using libquil as a QVM client
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Error when calling libquil_sys::quilc
+    #[error("error when calling libquil_sys: {0}")]
+    Quilc(#[from] libquil_sys::quilc::Error),
+    /// Error when serializing a program
+    #[error("error when serializing program: {0}")]
+    SerializeProgram(#[from] serde_json::Error),
+    /// Error when parsing a program
+    #[error("error when parsing program: {0}")]
+    ParseProgram(#[from] quil_rs::program::ProgramError),
+    /// Error when casting u64 to u32
+    #[error("error when casting u64 to u32: {0}")]
+    U64Truncation(#[from] TryFromIntError),
+    /// Error when creating a CString
+    #[error("error when creating CString: {0}")]
+    CString(#[from] NulError),
+}
+
+impl From<Error> for quilc::Error {
+    fn from(error: Error) -> Self {
+        quilc::Error::QuilcCompilation(error.to_string())
+    }
+}
+
+impl From<libquil_sys::quilc::CompilationMetadata> for NativeQuilMetadata {
+    fn from(value: libquil_sys::quilc::CompilationMetadata) -> Self {
+        NativeQuilMetadata {
+            final_rewiring: value
+                .final_rewiring
+                .iter()
+                .map(|r| u64::from(*r))
+                .collect::<Vec<_>>(),
+            gate_depth: value.gate_depth.map(u64::from),
+            gate_volume: value.gate_volume.map(u64::from),
+            multiqubit_gate_depth: value.multiqubit_gate_depth.map(u64::from),
+            program_duration: value.program_duration,
+            program_fidelity: value.program_fidelity,
+            topological_swaps: value.topological_swaps.map(u64::from),
+            qpu_runtime_estimation: value.qpu_runtime_estimation,
+        }
+    }
+}
+
+/// A libquil client providing Quilc functionality
+#[derive(Debug, Clone, Copy)]
+pub struct Client;
+
+impl quilc::Client for Client {
+    fn compile_program(
+        &self,
+        quil: &str,
+        isa: quilc::TargetDevice,
+        options: quilc::CompilerOpts,
+    ) -> Result<quilc::CompilationResult, quilc::Error> {
+        let program = libquil_sys::quilc::Program::from_str(quil).map_err(Error::from)?;
+        let isa = serde_json::to_string(&isa).map_err(Error::from)?;
+        let chip = libquil_sys::quilc::Chip::from_str(&isa).map_err(Error::from)?;
```

Comment:
I might be wrong, but this looks necessary, since it's going from a libquil error -> `Error` -> `super::quilc::Error`

Reply:
This could be made unnecessary by implementing `From<libquil::Error>` for `super::quilc::Error`, which would remove the need for this intermediate transformation by moving it into the `From` impl.

Reply:
It would actually have to be `From<libquil_sys::quilc::Error for super::quilc::Error`, and I don't think it makes sense to have `libquil_sys`-specific error handling in the parent module (`super::quilc`).

Reply:
Is this thread resolved? Looks like the `From<Error> for quilc::Error` above addresses this

Reply:
yes